### PR TITLE
Add support for in-memory uploads/downloads

### PR DESCRIPTION
**Pull request by m-cat**
# PULL REQUEST

## Overview

- Add `uploadData` and `downloadData`
- Add example/test scripts that use these methods.

## Checklist
Review and complete the checklist to ensure that the PR is complete before assigned to an approver.
 - [x] All new methods or updated methods have clear docstrings
 - [x] Testing added or updated for new methods
 - [ ] Verify if any changes impact the WebPortal Health Checks
 - [x] Approriate documentation updated
 - [x] Changelog file created

### CODE REVIEW

File: src/upload.js [hunk 1]
```diff
@@ -35,25 +35,45 @@ const defaultUploadOptions = {
...
+ * @param {string} filename - The filename to use on Skynet.
+ * @param {Object} [customOptions={}] - Configuration options.
+ * @returns - The skylink.
+ */
+SkynetClient.prototype.uploadData = async function (data, filename, customOptions = {}) {
+  const opts = { ...defaultUploadOptions, ...this.customOptions, ...customOptions };
+
+  const sizeInBytes = data.length;
+
+  if (sizeInBytes < opts.largeFileSize) {
```
Comment (peterjan) [inline src/upload.js:51]: this is kind of weird right, I can manipulate this through `customOptions`, shouldn't this be a constant?

**Review by peterjan — APPROVED** [review]
LGTM, what are we going to use this for exactly?

**Review by ro-tex — APPROVED** [review]

**Review by kwypchlo — APPROVED** [review]

**Comment by m-cat** [comment]
@peterjan This was originally requested by @dghelm for GunDB stuff, and lately requested again by a user. It's also useful for [this PR](https://github.com/SkynetLabs/nodejs-skynet/pull/140).

File: src/upload.js (hunk 1 above)
Comment (m-cat) [reply src/upload.js:51]: That's intended, below the threshold we use a single request and avoid the overhead of resumable uploads. The threshold is configurable by the dev.

**Comment by parajbs** [comment]
I also built a test skydb.js for "scripts" to check SkydB and SkydBV2.

**Comment by m-cat** [comment]
> I also built a test skydb.js for "scripts" to check SkydB and SkydBV2.

Great! Push it to your PR and I'll take a look.

**Comment by m-cat** [comment]
> Hello **all developers**,
> 
> I have work on SkyDB and SkyDBV2. I have already finished SkyDB, but I still have a problem with SkyDBV2. Maybe you can help?
> 
> How can you transform a skylink (see below) into a "rawDataLink" (see below)? This function (see below under SkyDB function) in "skynet-js" in "src/skydb_v2.ts" used in the lines 566-569!
> 
> **skylink:** sia://AADlJtYMBIjdjS0ovjLMkrfkIegApofn3yiGoyTPl15jfA
> 
> **trimuriprefix_dataLink:** AADlJtYMBIjdjS0ovjLMkrfkIegApofn3yiGoyTPl15jfA
> 
> **rawDataLink:** 0.0,229,38,21,136,221,45,40,50,204,146,183,232,0,135,233,163,207,94,99,124
> 
> rawDataLink must be a unit8array with 34 bits.
> 
> **skynet-Js: SkyDB functions:** const dataLink = trimUriPrefix(skyfile.skylink, URI_SKYNET_PREFIX); const rawDataLink = decodeSkylinkBase64(dataLink); validateUint8ArrayLen("rawDataLink", rawDataLink, "skylink byte array", RAW_SKYLINK_SIZE);
> 
> So far I could only turn back from rawDataLink into a skylink, but not to be turned around.
> 
> How can I get this, after that SkyDBV2 should also work. And then I can upload everything to github.
> 
> Please help.

Have you looked at the source code for that function?

```ts
export function decodeSkylinkBase64(skylink: string): Uint8Array {
  validateStringLen("skylink", skylink, "parameter", BASE64_ENCODED_SKYLINK_SIZE);
  // Add padding.
  skylink = `${skylink}==`;
  // Convert from URL encoding.
  skylink = skylink.replace(/-/g, "+").replace(/_/g, "/");
  return toByteArray(skylink);
}
```

Note that you'll also need the `base64-js` dependency:

```ts
import { fromByteArray, toByteArray } from "base64-js";
```

Also, please try to leave comments on the relevant PR so that things don't get lost. This PR has already been merged and closed.

**Pull request merged**